### DISCUSSION TITLE
Merge cli flag modules in vlib to one

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#11752] [Ideas]
We have `flag`, `cli` and `os.cmdline`. All of them should be merged to one module.

https://discord.com/channels/592103645835821068/592320321995014154/789781987338485781

**Comment by dedesite** [comment]
Hi,
With @timbasel we try to address this issue.
There is a PR open only for discussion here : https://github.com/timbasel/v/pull/1
I also created a gist which reflect how I see this "future" cli/flag module : https://gist.github.com/dedesite/1f1e26172729b022ef7833c934f8f0d8

I think before going forward, we need to be clear on what we want.

There has been discussion on discord with several contributor and here is the points were we seems to have found some sort of consensus : 

- the module should accept `--` and `-` for long flag syntax (ie `-prod`) but can encourage the first approach
- we can for example let the user create flag this way : `name: "prod"` and it will automatically become `--prod`.
- if the user want to have a single dash it could define the flag with : `name: "-prod"`
- Single letter flag should always be with a single dash
- No other flag syntax (like the DOS "/") should be allowed as even [Microsoft doesn't recommend it](https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-powershell-1.0/ee156811(v=technet.10)?redirectedfrom=MSDN)
- they may be declared with an `abbrev` field or maybe as a array of name (ie `name: ["prod", "p"])
- The `flag` API is really neat because you can declare flag and retrieve their value in the same time : 
`prod := fp.bool('prod', 0, false, 'Compile with optimisations.')`
- `cli` declarative style "makes the structure of the cli inherently clearer" and  separates the declaration from the parsing, but add several steps for retrieving flag values.
- we should have a way to define custom validator like eg: 
```v
	percent := git.get_float(cli.Flaf{
		name: 'percent'
		type_name: 'percent'
		flag: .custom,
		 validator: fn (value string) ?f64 {
			f_val := value.f64()
			if f_val >= 0 && f_val <= 100 { return f_val
			 } else { return error('Must be between 0 and 100')
			 }
		}
	})
```
- single letter boolean flag can be combine eg : `rm -rf` == `rm -r -f`
- flag can either be before argument or after. eg : `v run test.v -prod` == `v run -prod test.v`

If I come up with other thing, I will edit this message.

What do you think about that ?

**Comment by trollkarlen** [comment]
Can i suggest using the standard "--" and not "-" for multi char arguments.
This should be used on v itself also.
//T

**Comment by JalonSolov** [comment]
Using -- is a GNU standard, and has only existed as a "standard" since 1990.  POSIX and many others that have existed for _much_ longer (since the 1960s) only use - for all options, no matter the length.

**Comment by danieldaeschle** [comment]
To me follwing the newer standard makes more sense. :)

**Comment by JalonSolov** [comment]
Newer is not always better.  And GNU is only 2 years newer than POSIX (1988), so not enough difference to matter.

**Comment by danieldaeschle** [comment]
Are there any arguments that one dash is better than two? Obviously most tools use two dashes and it's cleaner (to me).

**Comment by JalonSolov** [comment]
Yes, most GNU tools, but not most tools. Many common tools that come preinstalled only use single dash long options, like `find`, and lots of other tools you can get with your package manager of choice also only use single dash.

The ONLY time it is useful to use two dashes is if there are short options with the same first letter as a long option... assuming there are any short options at all.  If not, it's just an extra character to type.

**Comment by trollkarlen** [comment]
Whats the take on this, no input since 2021  ?

**Reply by medvednikov** [reply]
Still needs to be done. Will do this year.

**Comment by daddycocoaman** [comment]
+1 for `--` as the standard for multi-character flags.

**Reply by daddycocoaman** [reply]
Sorry. Documentation wasn't clear that this is what ``posix_mode`` on Command would do! Just found out.
